File: src/main/java/me/beanbag/nuker/module/modules/canaltools/GenericProtectionLocator.java
package me.beanbag.nuker.module.modules.canaltools;

import me.beanbag.nuker.ModConfigs;
import net.minecraft.block.AbstractBlock;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.util.math.BlockPos;

import java.util.List;

public class GenericProtectionLocator implements IProtectionLocator {
    final static int underwaterMaxY = CanalSpecs.walkwayY - 1;
    final static int underwaterMinY = CanalSpecs.floorY + 1;
    final static int protectionMinX = CanalSpecs.westWallX + 1;
    final static int protectionMaxX = CanalSpecs.eastWallX - 1;

    final static int bridgeMinY = CanalSpecs.walkwayY;
    final static int bridgeMaxY = CanalSpecs.walkwayY + 3;
    @Override
    public boolean isProtection(int x, int y, int z) {
        return x >= protectionMinX && x <= protectionMaxX
                && (y >= underwaterMinY && y <= underwaterMaxY || y >= bridgeMinY && y <= bridgeMaxY);
    }

    @Override
    public boolean isCorrectInPosition(int x, int y, int z) {

        return statesForPosition(x, y, z).stream().map(AbstractBlock.AbstractBlockState::getBlock).toList().contains(
                ModConfigs.INSTANCE.getMc().world.getBlockState(new BlockPos(x, y, z)).getBlock());
    }

    @Override
    public List<BlockState> statesForPosition(int x, int y, int z) {
        if (!isProtection(x, y, z)) return List.of();
        if (y == underwaterMinY) {
            return List.of(Blocks.OBSIDIAN.getDefaultState());
        } else if (y == underwaterMaxY) {
            return  List.of(Blocks.OBSIDIAN.getDefaultState());
        } else if (y == bridgeMinY) {
            if (x > CanalSpecs.westWallX + 2 && x < CanalSpecs.eastWallX - 2) {
                return List.of();
            } else {
                return List.of(Blocks.AIR.getDefaultState());
            }
        } else if (y == bridgeMinY + 1) {
            if (x == protectionMinX + 1 || x == protectionMinX + 2 || x == protectionMaxX - 1 || x == protectionMaxX - 2) {
                return List.of(Blocks.OBSIDIAN.getDefaultState());
            } else {
                return List.of(Blocks.AIR.getDefaultState());
            }
        } else if (y == bridgeMinY + 2) {
            if (x <= protectionMaxX - 1 && x >= protectionMinX + 1) {
                if (x == protectionMaxX - 2 || x == protectionMinX + 2) {
                    return List.of(Blocks.OBSIDIAN.getDefaultState());
                } else {
                    return List.of(Blocks.OBSIDIAN.getDefaultState(), Blocks.CRYING_OBSIDIAN.getDefaultState(), Blocks.COBBLESTONE.getDefaultState());
                }
            } else {
                return List.of(Blocks.AIR.getDefaultState());
            }
        } else if (y == bridgeMinY + 3) {
            if (x <= protectionMaxX - 2 && x >= protectionMinX + 2) {
                if (x == protectionMaxX - 3 || x == protectionMinX + 3) {
                    return List.of(Blocks.OBSIDIAN.getDefaultState());
                } else {
                    return List.of(Blocks.OBSIDIAN.getDefaultState(), Blocks.CRYING_OBSIDIAN.getDefaultState());
                }
            } else {
                return List.of(Blocks.AIR.getDefaultState());
            }
        }
        return null;
    }
}

File: src/main/kotlin/mc/merge/module/modules/canal/tools/GenericProtectionLocator.kt
package mc.merge.module.modules.canal.tools

import mc.merge.ModCore
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.util.math.BlockPos


class GenericProtectionLocator : IProtectionLocator {
    override fun isProtection(x: Int, y: Int, z: Int): Boolean {
        return x in PROTECTION_MIN_X..PROTECTION_MAX_X && (y in UNDERWATER_MIN_Y..UNDERWATER_MAX_Y || y in BRIDGE_MIN_Y..BRIDGE_MAX_Y)
    }

    override fun isCorrectInPosition(x: Int, y: Int, z: Int): Boolean {
        return statesForPosition(x, y, z)!!.map { it.block }.toList().contains(
            ModCore.mc.world?.getBlockState(BlockPos(x, y, z))?.block
        )
    }

    override fun statesForPosition(x: Int, y: Int, z: Int): List<BlockState>? {
        if (!isProtection(x, y, z)) return listOf()
        when (y) {
            UNDERWATER_MIN_Y -> {
                return listOf(Blocks.OBSIDIAN.defaultState)
            }
            UNDERWATER_MAX_Y -> {
                return listOf(Blocks.OBSIDIAN.defaultState)
            }
            BRIDGE_MIN_Y -> {
                return if (x > CanalSpecs.WEST_WALL_X + 2 && x < CanalSpecs.EAST_WALL_X - 2) {
                    listOf()
                } else {
                    listOf(Blocks.AIR.defaultState)
                }
            }
            BRIDGE_MIN_Y + 1 -> {
                return if (x == PROTECTION_MIN_X + 1 || x == PROTECTION_MIN_X + 2 || x == PROTECTION_MAX_X - 1 || x == PROTECTION_MAX_X - 2) {
                    listOf(Blocks.OBSIDIAN.defaultState)
                } else {
                    listOf(Blocks.AIR.defaultState)
                }
            }
            BRIDGE_MIN_Y + 2 -> {
                return if (x <= PROTECTION_MAX_X - 1 && x >= PROTECTION_MIN_X + 1) {
                    if (x == PROTECTION_MAX_X - 2 || x == PROTECTION_MIN_X + 2) {
                        listOf(Blocks.OBSIDIAN.defaultState)
                    } else {
                        listOf(
                            Blocks.OBSIDIAN.defaultState,
                            Blocks.CRYING_OBSIDIAN.defaultState,
                            Blocks.COBBLESTONE.defaultState
                        )
                    }
                } else {
                    listOf(Blocks.AIR.defaultState)
                }
            }
            BRIDGE_MIN_Y + 3 -> {
                return if (x <= PROTECTION_MAX_X - 2 && x >= PROTECTION_MIN_X + 2) {
                    if (x == PROTECTION_MAX_X - 3 || x == PROTECTION_MIN_X + 3) {
                        listOf(Blocks.OBSIDIAN.defaultState)
                    } else {
                        listOf(Blocks.OBSIDIAN.defaultState, Blocks.CRYING_OBSIDIAN.defaultState)
                    }
                } else {
                    listOf(Blocks.AIR.defaultState)
                }
            }
            else -> return null
        }
    }

    companion object {
        const val UNDERWATER_MAX_Y: Int = CanalSpecs.WALKWAY_Y - 1
        const val UNDERWATER_MIN_Y: Int = CanalSpecs.FLOOR_Y + 1
        const val PROTECTION_MIN_X: Int = CanalSpecs.WEST_WALL_X + 1
        const val PROTECTION_MAX_X: Int = CanalSpecs.EAST_WALL_X - 1

        const val BRIDGE_MIN_Y: Int = CanalSpecs.WALKWAY_Y
        const val BRIDGE_MAX_Y: Int = CanalSpecs.WALKWAY_Y + 3
    }
}
